src/python/deepgnn/graph_engine/snark/decoders.py

```diff
@@ -43,6 +43,191 @@ def decode(self, line: str) -> Iterator[Tuple[int, int, int, float, list]]:
+        0 -1 1 1 1 1.1 1.1
+        0 1 0 4 10 1 1 1
+        1 -1 1 1 1 1.1 1.1
+        1 0 0 4 10 1 1 1
+        ```
+    """
+
+    def __init__(
+        self,
+        default_node_type: Optional[int] = None,
+        default_edge_type: Optional[int] = None,
+        default_node_weight: Optional[float] = None,
+        default_edge_weight: Optional[float] = None,
+        default_node_feature_types: Optional[List[Optional[str]]] = None,
+        default_edge_feature_types: Optional[List[Optional[str]]] = None,
+        default_node_feature_lens: Optional[List[Optional[List[int]]]] = None,
+        default_edge_feature_lens: Optional[List[Optional[List[int]]]] = None,
+    ):
+        """Initialize the Decoder."""
+        super().__init__()
+        self.node_type = default_node_type
+        self.node_weight = default_node_weight
+        self.node_feature_types = (
+            default_node_feature_types if default_node_feature_types is not None else []
+        )
+        self.node_feature_lens = (
+            default_node_feature_lens if default_node_feature_lens is not None else []
+        )
+        self.n_node_feature = len(self.node_feature_types)
+
+        self.edge_type = default_edge_type
+        self.edge_weight = default_edge_weight
+        self.edge_feature_types = (
+            default_edge_feature_types if default_edge_feature_types is not None else []
+        )
+        self.edge_feature_lens = (
+            default_edge_feature_lens if default_edge_feature_lens is not None else []
+        )
+        self.n_edge_feature = len(self.edge_feature_types)
+
+    def decode(self, line: str) -> Iterator[Tuple[int, int, int, float, list]]:
+        """Convert text line into a node and edge iterator.
+
+        This is a generator that yields a node then its outgoing edges in order.
+        A node and its outgoing edges can be on different partitions.
+        Yield format is (node_id/src, -1/dst, type, weight, features).
+        Features being a list of dense features as ndarrays and sparse features as 2 tuples, coordinates and values.
+        """
+        if line == "":
+            return []
+
+        data = iter(line.split(","))
+
+        src, dst = int(next(data)), int(next(data))
+        if dst == -1:
+            typ, weight = self.node_type, self.node_weight
+            item_feature_types, item_feature_lens, default_feature_len = (
+                self.node_feature_types,
+                self.node_feature_lens,
+                self.n_node_feature,
+            )
+        else:
+            typ, weight = self.edge_type, self.edge_weight
+            item_feature_types, item_feature_lens, default_feature_len = (
+                self.edge_feature_types,
+                self.edge_feature_lens,
+                self.n_edge_feature,
+            )
+        if typ is None:
+            typ = int(next(data))
+        if weight is None:
+            weight = float(next(data))
+
+        features = []
```

Comment:
could you extract feature parsing in a separate function? It is hard to follow through.